src/zino/statemodels.py

```diff
@@ -363,3 +368,156 @@ class AlarmEvent(Event):
+        for event in events:
+            # get special handling of embryonic -> open transition first
+            state.events.commit(event)
+            event.state = EventState.IGNORED
+            state.events.commit(event)
+            self.event_ids.append(event.id)
+
+    def end(self, state: "ZinoState"):
+        for event_id in self.event_ids:
+            event = state.events.checkout(event_id)
+            event.state = EventState.OPEN
+            state.events.commit(event)
+
+    def add_log(self, message: str) -> LogEntry:
+        entry = LogEntry(message=message)
+        self.log.append(entry)
+        return entry
+
+    def matches_event(self, event: Event, state: "ZinoState") -> bool:
+        """Returns true if `event` will be affected by this planned maintenance"""
+        raise NotImplementedError
+
+    def _get_or_create_events(self, state: "ZinoState") -> list[Event]:
+        """Creates/gets events that are affected by the given starting planned
+        maintenance
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def _string_match(pattern: str, string: str) -> bool:
+        """This should behave like tcl string match https://wiki.tcl-lang.org/page/string+match
+        Returns true if `string` matches `pattern`.
+        """
+        return fnmatch.fnmatch(string, pattern)
+
+    @staticmethod
+    def _regex_match(pattern: str, string: str) -> bool:
+        """Matches `string` against regex expression `pattern`.
+        Returns true if there is a match.
+        """
+        compiled_pattern = re.compile(pattern)
+        return bool(compiled_pattern.match(string))
+
+
+class DeviceMaintenance(PlannedMaintenance):
+    type: PmType = PmType.DEVICE
+
+    def matches_event(self, event: Event, state: "ZinoState") -> bool:
+        """Returns true if `event` will be affected by this planned maintenance"""
+        if event.type not in ["reachability", "alarm"]:
+            return False
+        device = state.devices[event.router]
+        return self.matches_device(device)
+
+    def matches_device(self, device: DeviceState) -> bool:
+        """Returns true if ReachabilityEvents and AlarmEvents related to `device`
+        would be affected by this planned maintenance
+        """
+        if self.match_type == "regexp":
+            return self._regex_match(self.match_expression, device.name)
+        if self.match_type == "str":
+            return self._string_match(self.match_expression, device.name)
+        if self.match_type == "exact":
+            return self.match_expression == device.name
+        return False
+
+    def _get_or_create_events(self, state: "ZinoState") -> list[Event]:
+        """Creates/gets events that are affected by the given starting planned
+        maintenance
+        """
+        events = []
+        # all devices that the pm should affect
+        devices = [device for device in state.devices.devices.values() if self.matches_device(device)]
```

Comment:
This doesn't need to keep a full list in memory, so it could simply use a generator expression here.
```suggestion
        devices = (device for device in state.devices.devices.values() if self.matches_device(device))
```

Reply:
fixed